docs/proposals/koordlet/20231227-koordlet-resctrl-qos-enhance.md

```diff
@@ -0,0 +1,535 @@
+    closid := GetClosId(annotation, label)
+    app, err := p.engine.GetApp(podid)
+    updateContainerContext(podid, containerid, app.closid, app.Hooks)
+}
+
+// UpdatePodTaskIds will update new taskids to resctrl file system
+func (p *plugin) UpdatePodTaskIds(proto protocol.HooksProtocol) error {
+    // 1. retrieve task ids for each slo and each specific LLC/MBA request pod by consume pod taskids
+    // 2. add the new related task ids in resctrl groups
+}
+```
+
+**PodContext**
+```go
+
+func (p *PodContext) NriDone(executor resourceexecutor.ResourceUpdateExecutor) {
+    if p.executor == nil {
+        p.executor = executor
+    }
+    p.injectForExt()
+    p.Update()
+}
+
+func (p *PodContext) NRIRemoveDone() {
+    if p.executor == nil {
+        p.executor = executor
+    }
+    p.removeForExt()
+    p.Update()
+}
+
+// removeForExt will handle NRI remove/clean operation
+func (p *PodContext) removeForExt() {
+    
+}
+func injectResctrl(closid string, schemata string, a *audit.EventHelper, e resourceexecutor.ResourceUpdateExecutor) (resourceexecutor.ResourceUpdater, error) {
+    // for specific pods, create control group and monitor group and update schemata
+}
+
+func removeResctrl(closid string, a *audit.EventHelper, e resourceexecutor.ResourceUpdateExecutor) (resourceexecutor.ResourceUpdater, error) {
+    // for specific pods, create control group and monitor group and update schemata
+}
+```
+
+**ContainerContext**
+```go
+func (c *ContainerContext) NriDone(executor resourceexecutor.ResourceUpdateExecutor) (*api.ContainerAdjustment, *api.ContainerUpdate, error) {
+    ......
+    if c.Response.Resources.Resctrl != nil {
+        // adjust OCI spec
+        adjust.SetLinuxRDTClass(c.Response.Resources.Resctrl.Closid)
+        adjust.Hooks = c.Response.Resources.Resctrl.Hooks
+    }
+}
+```
+
+**Resources**
+
+```go
+type Resctrl struct {
+    Schemata string
+    Hooks string
+    Closid string
+}
+
+type Resources struct {
+    // origin resources
+    CPUShares   *int64
+    CFSQuota    *int64
+    CPUSet      *string
+    MemoryLimit *int64
+
+    // extended resources
+    CPUBvt *int64
+    Resctrl *Resctrl
+}
+```
+
+**ResctrlUpdater**
+
+We will enhance current ResctrlUpdate to support directly overwrite schemata. Current ResctrlUpdater doesn't support override schemata with LLC and MBA together directly and doesn't consider NUMA when apply schemata .
+
+```go
+func NewResctrlSchemata(group, schemata string) ResourceUpdater {
+
+}
+
+func NewResctrlSchemataWithNuma(group, resctrl Resctrl) ResourceUpdater {
+
+}
+
+```
+
+**podsInformer**
+
+podsInformer will add a new interface to get all pods' taskids, Resctl runtime hook Reconciler will register callback to consume these information and write new taskIds to resctrl contrl group and monitor group.
+
+```go
+
+type PodMeta struct {
+    Pod       *corev1.Pod
+    CgroupDir string
+    // Add new field, todo: memory usage
+    ContainerTaskIds map[string][]uint64
```

Comment:
As for the memory overhead, since the pod meta can be frequently copied and sent to any runtime hook or QoS plugin even if they do not use the IDs, we may need a new informer to maintain the container task IDs. Or we can add an option to determine whether this `taskIDs` field is maintained. WDYT

Reply:
Yes, as you said, we will add an option currently. In the future, we may consider adding a new informer which contains task IDs and other similar information.

Reply:
Perhaps `--enable-pod-task-ids-informer` as a cmd line arg and a boolean field in the `Config`?

Reply:
Anyway, I think some details can be discussed further in the implementation PRs.

Reply:
> Perhaps `--enable-pod-task-ids-informer` as a cmd line arg and a boolean field in the `Config`?

OK, let's discuss about it in the implementation PRs